docs/PSEUDOCODE.md

```diff
@@ -0,0 +1,191 @@
+        if upkeep is retryable now
+            add to next retries
+
+send retries as tick
+```
+
+### Conditional Sampling
+To reduce over all check load on the nodes for conditional upkeeps, each node
+only checks a random subset of upkeeps per tick interval. The tick interval is
+derived by block cadence, but not every chain will have every block checked.
+Fast chains will have multiple blocks skipped while sampling takes place.
+```
+
+```
+
+## OCR3 Report Flow
+OCR is allowed to run on a separate cadence from the eligibility check flow. In
+general, it pulls from some collection of staged values, constructs observations,
+and builds reports. Some components hook into this process to feed progress
+data to the eligibility check flows.
+
+For the case of node restarts, it was proposed that local caches would be
+cleared exposing a possibility that upkeeps could be reported by a set of 
+restarted nodes resulting in successful (bad) transmits.
+
+Approach #1: maintain recent results in outcome and use as a filter for new
+outcomes. (not shown)
+
+Approach #2: consider the lack of reporting an upkeep as a valid observation.
+f+1 nodes can indicate that an upkeep needs to be performed while f+1 nodes can
+also indicate that an upkeep does not need to be performed. In this case, both
+observations would achieve a valid quorum, but produce different outcomes. 
+Comparing the observations against each other, accounting for both the positive 
+*AND* negative, a final unified outcome can be achieved that is resistant to
+node restarts or fresh nodes entering the network. (shown)
+```
+for each round
+    // Observation -------------------
+    for each check pipeline result in previous outcome
+        add check pipeline result to in-flight cache
+        remove check pipeline result from ocr staging
+
+    // filtering can be applied within ocr staging such that the previous add
+    // to in-flight is immediately applied to results being pulled
+    get check pipeline results from ocr staging
+    
+    for each check pipeline result
+        add check pipeline result to observation
+
+    get latest block from sample staging
+    add latest block to observation
+
+    // only eligible upkeeps
+    get sampled upkeep ids from sample staging
+
+    for each sampled upkeep id
+        add upkeep id to observation
+
+    // observation contains: latest block, sampled ids, check pipeline results
+    return observation
+
+    // Outcome -------------------
+    for each attributedObservation
+        increment count for check pipeline result
+        update median block number with observation latest
+        add observation id samples to sampled id list
+
+    for each attributedObservation
+        for each flattened check pipeline result
+            if check pipeline result is not in observation
+                decrement count for check pipeline result
+
+    for each flattened check pipeline result
+        if count is greater than 0
```

Comment:
I assume this is a proxy for f+1 quorum? Does this depend on the number of observations a leader gets?